data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +230,219 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
+                            valueEnd = i;
+                            bracketStack.pop();
+                            valueString = pair.substring(valueStart, valueEnd).stripLeading();
+                            JsonNode child = ((ObjectNode) root).put(keyString, recurse(valueString, mapper));
+                        }
+                    }
+                }  
+            } else {
+                valueString = pair.substring(valueStart).stripLeading();
+                ObjectNode child = ((ObjectNode)root).put(keyString, valueString);
+            }
+        }
+
+        return root;
+    }
+
+    private static Map<Character, Character> initBracketMap() {
+        Map<Character, Character> bracketMap = new HashMap<>();
+
+        bracketMap.put('[', ']');
+        bracketMap.put('(', ')');
+        bracketMap.put('<', '>');
+
+        return bracketMap;
+    }
+
+    private Map<String, Object> createRecursedMap(JsonNode node, ObjectMapper mapper) {
+        return mapper.convertValue(node, new TypeReference<HashMap<String, Object>>() {});
+    }
+
+    private Map<String, Object> createNonRecursedMap(String[] groups) {
+        Map<String, Object> nonRecursedMap = new LinkedHashMap<>();
+        List<Object> valueList;
+
+        for(final String group : groups) {
+            final String[] terms = keyValueDelimiterPattern.split(group, 2);
+            String key = terms[0];
+            Object value;
+            
+            if (terms.length == 2) {
+                value = terms[1];
+            } else {
+                LOG.debug("Unsuccessful match: '{}'", terms[0]);
+                value = keyValueProcessorConfig.getNonMatchValue();
+            }
+
+            if (nonRecursedMap.containsKey(key)) {
+                Object existingValue = nonRecursedMap.get(key);
+
+                if (existingValue instanceof List) {
+                    valueList = (List<Object>) existingValue;
+                } else {
+                    valueList = new ArrayList<Object>();
+                    valueList.add(existingValue);
+                    nonRecursedMap.put(key, valueList);
                 }
 
-                if (keyValueProcessorConfig.getRemoveBrackets()) {
-                    final String bracketRegex = "[\\[\\]()<>]";
-                    if (value != null) {
-                        value = value.toString().replaceAll(bracketRegex,"");
+                if (keyValueProcessorConfig.getSkipDuplicateValues()) {
+                    if (!valueList.contains(value)) {
+                        valueList.add(value);
                     }
+                } else {
+                    valueList.add(value);
                 }
+            } else {
+                nonRecursedMap.put(key, value);
+            }
+        }
 
-                addKeyValueToMap(parsedMap, key, value);
+        return nonRecursedMap;
+    }
+
+    private void executeConfigs(Map<String, Object> map, Map<String, Object> parsed) {
```

Comment:
As mentioned above, parsed can be removed from input args and become the output of the method